File: backend/internal/model/pet.go
package model

import (
	"time"
)

type Pet struct {
	ID            string     `json:"id"`
	OwnerID       string     `json:"owner_id"`
	Name          string     `json:"name"`
	Type          string     `json:"type"`
	Breed         string     `json:"breed"`
	Age           *int       `json:"age"`
	LostDate      *time.Time `json:"lost_date"`
	Description   string     `json:"description"`
	ExternalSigns string     `json:"external_signs"`
	ContactPhone  *string    `json:"contact_phone"`
	ContactName   *string    `json:"contact_name"`
	ImageURLs     []string   `json:"image_urls"`
	Latitude      *float64   `json:"latitude"`
	Longitude     *float64   `json:"longitude"`
	Address       string     `json:"address"`
	Status        string     `json:"status"`
	CreatedAt     time.Time  `json:"created_at"`
	UpdatedAt     time.Time  `json:"updated_at"`
}

type CreatePetRequest struct {
	Name          string    `json:"name" binding:"required"`
	Type          string    `json:"type" binding:"required"`
	Breed         string    `json:"breed"`
	Age           *int      `json:"age"`
	LostDate      time.Time `json:"lost_date" binding:"required"`
	Description   string    `json:"description"`
	ExternalSigns string    `json:"external_signs"`
	ContactPhone  string    `json:"contact_phone"`
	ContactName   string    `json:"contact_name"`
	ImageURLs     []string  `json:"image_urls"`
	Latitude      float64   `json:"latitude" binding:"required"`
	Longitude     float64   `json:"longitude" binding:"required"`
	Address       string    `json:"address" binding:"required"`
}

type ListPetsRequest struct {
	Latitude   float64 `form:"lat" binding:"required"`
	Longitude  float64 `form:"long" binding:"required"`
	Radius     float64 `form:"radius" binding:"required"`
	SearchText string  `form:"search"`
	PetType    string  `form:"type"`
	LostFrom   string  `form:"lost_from"`
	LostTo     string  `form:"lost_to"`
	Breed      string  `form:"breed"`
}

type Sighting struct {
	ID             string    `json:"id"`
	PetID          string    `json:"pet_id"`
	ReporterID     string    `json:"reporter_id"`
	ReporterName   string    `json:"reporter_name"`
	ReporterAvatar string    `json:"reporter_avatar"`
	Latitude       float64   `json:"latitude"`
	Longitude      float64   `json:"longitude"`
	Address        string    `json:"address"`
	Comment        string    `json:"comment"`
	ImageURLs      []string  `json:"image_urls"`
	Status         string    `json:"status"`
	CreatedAt      time.Time `json:"created_at"`
}

File: PetFinderApp/PetFinderApp/Models/Pet.swift
//
//  Pet.swift
//  PetFinderApp
//
//  Created by Вилина Ольховская on 26.08.2025.
//
import Foundation

struct Pet: Identifiable, Codable, Hashable {
    let id: String
    let createdAt: Date?
    let name: String
    let breed: String
    let description: String
    let location: String
    let imageUrls: [String]
    let age: Int?
    let externalSigns: String?
    let longitude: Double?
    let latitude: Double?
    let type: String
    let contactPhone: String?
    let contactName: String?
    let lostDate: Date?
    let ownerId: String?
    let status: String?

    enum CodingKeys: String, CodingKey {
        case id, name, breed, description, age, latitude, longitude, type, status
        case location = "address"   
        case createdAt
        case imageUrls
        case externalSigns
        case contactPhone
        case contactName
        case lostDate
        case ownerId
    }

    var daysLost: Int? {
        guard let date = lostDate ?? createdAt else { return nil }
        let days = Calendar.current.dateComponents([.day], from: date, to: Date()).day ?? 0
        return max(0, days)
    }

    var timeAgo: String {
        guard let days = daysLost else { return "" }
        if days == 0 { return "сегодня" }
        return "\(days) дн"
    }
}

// MARK: - Sighting

struct Sighting: Codable {
    let id: String
    let petId: String
    let reporterId: String
    let reporterName: String
    let reporterAvatar: String
    let latitude: Double
    let longitude: Double
    let address: String
    let comment: String
    let imageUrls: [String]
    var status: String
    let createdAt: Date?
}
